1760 cache spin and repoint dependencies in pre processing

tmplummer opened the pull request:
# Change Summary

## Overview
This PR adds automatic handling of spin and repoint dependencies in the `pre_processing` step. Rather than loading tables from an environment variable, mutable module level attributes are set up to hold the path or paths to the CSV files. When the `get_spin_data()` or `get_repoint_data()` functions are called, the path(s) are pulled from those attributes.

Additionally, I added caching of the dataframes produced so that subsequent calls don't require reloading the table from disk.

## Updated Files
- imap_processing/cli.py
   - Set the spin and repoint paths in the pre_processing step
- imap_processing/spice/repoint.py
   - Add mutable module level attribute for storing the path to the repoint table
   - Add setter for the above attribute
   - Use the new attribute when loading the repoint table from disk
   - Add caching of the repoint table to the `get_repoint_data()` function
- imap_processing/spice/spin.py
   - Add mutable module level attribute for storing the path to the spin tables
   - Add setter for the above attribute
   - Use the new attribute when loading the spin tables from disk
   - Add caching of the spin table dataframe to the `get_spin_data()` function
   - Add merging of multiple input spin tables
- imap_processing/tests/conftest.py
   - Fix up spin and repoint table fixtures to set module global attributes instead of environment variables
- imap_processing/tests/spice/test_repoint.py
   - Update tests for repoint changes
   - Add test coverage for attribute setter
- imap_processing/tests/spice/test_spin.py
   - Update tests for spin changes
   - Add test coverage for attribute setter
   - Add test for merging multiple spin tables
   - 
- imap_processing/tests/test_cli.py
   - Add test coverage for setting spin and repoint paths from dependencies

Closes: #1760

tmplummer committed:
Set paths to spin and repoint files in pre_processing
Add functionality to merge multiple spin tables
Load spin and repoint tables from mutable module global attribute

tmplummer committed:
fix existing spin and repoint tests

tmplummer committed:
Fix interdependence of spin and repoint tests by using monkeypatch
Add test coverage for setting of spin and repoint table paths in pre_processing

tmplummer committed:
Implement my own cache functionality to avoid issues with using global

tmplummer committed:
Add autouse fixture that provides clean paths in spin and repoint modules

maxinelasp commented on imap_processing/spice/repoint.py:
```diff
@@ -1,16 +1,49 @@
 logger = logging.getLogger(__name__)
 
+# Use a mutable module level attribute to store the location of the repoint table
+_repoint_table_path: Path | None = None
```
I really don't like the global variable... why not stick this stuff into a class so we can pass around an instance? You could also put it in the `SpinInput` class (as some kind of set_dataframe() method) and then it will operate in the same kind of way as this does. 

If you don't want to update imap_data_access you could extend ProcessingInput into an imap_processing specific class and replace the dependency with a new ProcessingInput with the dataframe included.

tmplummer replied:
I generally agree that global variables should be avoided. I think this is an acceptable time to use them. I view the spin table query functions as being very similar to the SPICE queries. I don't think we want to have to pass spin and repoint paths, or dataframe objects, through many layers of functions in order to make the data available where it is needed. SPICE works in a very similar way where we tell the SPICE library about the data in pre-processing and then make queries throughout our code to retrieve what we want.

maxinelasp replied:
We're already passing around ancillary files and I think this falls in the same category - ie it is a data source that code needs to access and so it should be explicitly passed around into places that use it. I dislike having implicit dependencies like this, even if it means we need to pass around more objects. I also think this design would be difficult to test against - i.e. not difficult to test, but difficult if you depend on it. 

I am also worried about the global object getting mutated and messing up processing at later steps. Those kinds of bugs can be really difficult to track down. If we could make the object an immutable type I would feel better about it. 

I took a peek into the Spiceypy code, and the `furnsh` method actually calls directly out to SPICE so I think that's a different case. 

Things that would make me feel better about this design:
- Putting the variable into a singleton class so we can name the class something like `SpinSingleton` and make it clear this is globally shared. I also worry about people by-passing the `get()` method you have and using the variable directly. You can then use the `@property` decorator to override the class method for `get()` and people can then use `SpinSingleton.spin` and always use the get() method you wrote.
- if the spin table path is ever updated, we should invalidate the cache

tmplummer replied:
This seems like an impasse. I simply don't have the time to totally re-architect this. Can global variables cause hard to debug issues? Sure. I think that the following mitigate the hazards sufficiently:
- The global is named starting with an underscore indicating it is private so no one should directly mutate it.
- Even if someone did change the spin or repoint files pointed to by the global, things would be OK.
   - The `get_spin_data()` function sorts the files to the appropriate order so that when they are combined, the resulting dataframe is correct. The only side effect would be that there might not be the coverage needed in which case an exception is raised saying that the query time is outside of the spin-table data.
   - Changing the repoint file will simply mean that the time coverage is not the same. Each repoint table file is the same as the prior version with additional data appended. Again, worst case, an exception gets raised.

> I took a peek into the Spiceypy code, and the furnsh method actually calls directly out to SPICE so I think that's a different case.

Sure, but just like someone can mutate a private module variable, someone can also call `spiceypy.furnish()` with some random kernel, or even a reordered list of the original kernels that were furnished, and royally screw things up.

## Path forward?
I guess that I will cherry pick the changes to the `get_spin_data()` class that add the functionality to combine multiple spin files and then make the current environment variable implementation carry a json string with the multiple file paths. It won't change any of your concerns, but perhaps we can get past SIT4 with it?

maxinelasp replied:
Well, I hate the way SPICE does it, but necessary evils!

Since we're under a tight deadline for SIT-4, I am ok with this being merged. Maybe we can re-visit after the SIT.

greglucas replied:
I think this is a place where we do want to somehow load these automagically for users without needing to pipe them through everywhere similar to SPICE, so I think this makes sense. Maybe a middle-ground would be to remove the global calls by moving these global items to a separate singleton module that can be referenced here instead? I think that would be similar to Maxine's suggestion of the singleton class, but without the need for defining a class and rather just defining your attribute at the module level instead of class level.

Something like this setup:

**spice/config.py**
```python
_repoint_table_path = None
_spin_table_paths = None
```

**spice/repoint.py**
```python
from imap_processing.spice import config

def set_repoint_path(path):
    config._repoint_table_path = path

...
```

maxinelasp commented on imap_processing/spice/spin.py:
```diff
@@ -13,14 +14,45 @@
+        List of paths to spin-table csv files that will be used to supply
+        spin-table data.
+    """
+    global _spin_table_paths  # noqa: PLW0603
```
I don't think we should `noqa` this particular ruff rule - I agree with ruff that global variables can cause a lot of difficult-to-understand issues

maxinelasp reviewed (changes requested):
I think these changes should actually probably be in imap-data-access - since we're passing around classes with dependencies in them, we can put the spin pandas dataframe in there. I can see these global variables causing lots of difficult to diagnose problems for us...

tmplummer committed:
Fix spin table combining logic
Add test coverage for combining of multiple spin tables

greglucas commented on imap_processing/spice/repoint.py:
```diff
@@ -1,16 +1,49 @@
+_repoint_df_cache: dict[str, pd.DataFrame] = {}
+
+
+def set_repoint_table_paths(paths: list[Path]) -> None:
```
Do we want to name these similar to SPICE so call it furnish_repoint_file() or something like that? Maybe set_global_repoint_table() to add in global and indicate it is updating this globally.

greglucas commented on imap_processing/spice/repoint.py:
```diff
@@ -42,16 +75,27 @@ def get_repoint_data() -> pd.DataFrame:
+        f"Reading in the following repoint table file: {_repoint_table_path.name}"
+    )
+
+    cache_key = str(_repoint_table_path.resolve())
+    if cache_key in _repoint_df_cache:
+        logger.debug("Returning cached repointing data")
+        return _repoint_df_cache[cache_key]
```
I don't think this will work equivalently to what you have here since there is some extra processing involved, but wanted to mention that sometimes I have made a separate function that can be cached internal to your external function so that I don't have to make a cache and can just leverage the functools implementation. Something like:

```python
def get_repoint_data():
    path = _from_global
    repoint_df = _load_repoint_data_with_cache(path)

@functools.cache
def _load_repoint_data_with_cache(path):
    return pd.read_csv(path)
```

tmplummer replied:
Good call. I was thinking along these lines, but couldn't quite make the leap to this pattern.

greglucas reviewed (approved):
I agree that this is something that will be helpful to get in for SIT-4, and hopefully any clean-up would only improve on the approach but wouldn't mean this was wasted effort / wrong. I have a few minor suggestions/comments, but they are non-critical and can be merged over if you don't have time/desire to look into them right now.

tmplummer committed:
Move global spin and repoint location config into new spice.config module
Extract spin and repoint reading code into private methods to allow for use of functools.cache

tmplummer committed:
Add named spin_number column that duplicates index so that it can be referred to by name

tmplummer committed:
Fix some tests

tmplummer committed:
Fix Docs build

lacoak21 committed:
IDEX pipeline fixes (#1773)

* pipeline fixes

* global names

* fix spin phase format

tmplummer committed:
1729 add pointing frame kernel job to clipy (#1776)

* Add high level function for producing pointing attitude kernel

* Add pointing kernel job to cli Spacecraft.do_processing
Create custom Spacecraft.post_processing to handle list[Path] returned by generate_poiting_attitude_kernel()

* bump imap-data-access version

* add poetry.lock updates

* Allow for Paths to be passed in to default post_processing method

* Fix bug found in integration testing

* Improve pointing kernel name generation

tmplummer committed:
Remove setting of pandas copy_on_write=True... it was causing seemingly unrelated failures

tmplummer committed:
empty commit to try to unhang github
